crates/loans/src/lib.rs

```diff
@@ -0,0 +1,2055 @@
+        type UpdateOrigin: EnsureOrigin<Self::Origin>;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+
+        /// Unix time
+        type UnixTime: UnixTime;
+
+        /// Assets for deposit/withdraw collateral assets to/from loans module
+        type Assets: Transfer<Self::AccountId, AssetId = CurrencyId, Balance = Balance>
+            + Inspect<Self::AccountId, AssetId = CurrencyId, Balance = Balance>
+            + Mutate<Self::AccountId, AssetId = CurrencyId, Balance = Balance>;
+
+        /// Reward asset id.
+        #[pallet::constant]
+        type RewardAssetId: Get<AssetIdOf<Self>>;
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Insufficient liquidity to borrow more or disable collateral
+        InsufficientLiquidity,
+        /// Insufficient deposit to redeem
+        InsufficientDeposit,
+        /// Repay amount greater than allowed
+        TooMuchRepay,
+        /// Asset already enabled/disabled collateral
+        DuplicateOperation,
+        /// No deposit asset
+        NoDeposit,
+        /// Repay amount more than collateral amount
+        InsufficientCollateral,
+        /// Liquidator is same as borrower
+        LiquidatorIsBorrower,
+        /// Deposits are not used as a collateral
+        DepositsAreNotCollateral,
+        /// Insufficient shortfall to repay
+        InsufficientShortfall,
+        /// Insufficient reserves
+        InsufficientReserves,
+        /// Invalid rate model params
+        InvalidRateModelParam,
+        /// Market not activated
+        MarketNotActivated,
+        /// Oracle price not ready
+        PriceOracleNotReady,
+        /// Oracle price is zero
+        PriceIsZero,
+        /// Invalid asset id
+        InvalidCurrencyId,
+        /// Invalid lend_token id
+        InvalidLendTokenId,
+        /// Market does not exist
+        MarketDoesNotExist,
+        /// Market already exists
+        MarketAlreadyExists,
+        /// New markets must have a pending state
+        NewMarketMustHavePendingState,
+        /// Upper bound of supplying is exceeded
+        SupplyCapacityExceeded,
+        /// Upper bound of borrowing is exceeded
+        BorrowCapacityExceeded,
+        /// Insufficient cash in the pool
+        InsufficientCash,
+        /// The factor should be greater than 0% and less than 100%
+        InvalidFactor,
+        /// The supply cap cannot be zero
+        InvalidSupplyCap,
+        /// The exchange rate should be greater than 0.02 and less than 1
+        InvalidExchangeRate,
+        /// Amount cannot be zero
+        InvalidAmount,
+        /// Payer cannot be signer
+        PayerIsSigner,
+        /// Codec error
+        CodecError,
+        /// Collateral is reserved and cannot be liquidated
+        CollateralReserved,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (crate) fn deposit_event)]
+    pub enum Event<T: Config> {
```

Comment:
I think we should switch to named events before this is live ideally to limit downstream (squid) changes

Reply:
Create issue here: https://github.com/interlay/interbtc/issues/757